bin/lib/cdn.py

```diff
@@ -0,0 +1,226 @@
+import hashlib
+import tarfile
+import urllib.parse
+
+from base64 import b64encode
+from concurrent import futures
+from datetime import datetime
+from tempfile import mkdtemp
+import os
+import shutil
+import logging
+from pathlib import Path
+
+from lib.amazon import botocore, s3_client, force_lazy_init
+
+logger = logging.getLogger('ce-cdn')
+
+def read_file_chunked(fobj, chunk_size=128*1024):
+    b = bytearray(chunk_size)
+    mv = memoryview(b)
+
+    for n in iter(lambda : fobj.readinto(mv), 0):
+        yield mv[:n]
+
+
+def hash_fileobj(fobj, hash_type):
+    h = hash_type()
+    for chnk in read_file_chunked(fobj):
+        h.update(chnk)
+    return h
+
+
+def hash_file_for_s3(f):
+    with open(f['path'], 'rb') as fobj:
+        sha256 = hash_fileobj(fobj, hashlib.sha256).digest()
+        sha256 = b64encode(sha256).decode()
+        return dict(hash=sha256, **f)
+
+
+def get_directory_contents(basedir):
+    for f in Path(basedir).rglob('*'):
+        name = f.relative_to(basedir).as_posix()
+        yield dict(name=name, path=f)
+
+
+class DeploymentJob(object):
+    tmpdir = None
```

Comment:
Is this a member variable? My python is weak here; does this not make `tmpdir` a class variable (e.g. a static?) if so doesn't that cause issues? (e.g. if there were two Jobs constructed at once?) Or am I missing something? :)